### PR TITLE
feat: Introduce new scene cache management

### LINKED ISSUE #38: [EgoVenture] Enhance scene cache by adding target scenes of hotspots
**Please make sure you talk to the community before creating an issue.**

Project:
- [X] Egoventure

**Is your feature request related to a problem? Please describe.**
EgoVenture's cache is depending on the numbering of scenes. However it is not always possible to connect the scenes in this way. Take for example Carol Refurbished where man05.tscn includes a walk hotspot to man14.tscn.

**Describe the solution you'd like**
I would propose to add the scenes to the cache that are listed in the hotspots of the scene passed to update_cache().

**Additional context**
Here's a small _prototype_ how such a feature could be added in scene_cache.gd. Such logic could get added in function update_cache() merging it with the removal/adding of scenes based on the index sequence.

```
func _add_linked_scenes(current_scene: String):
	var scene_node = get_scene(current_scene).instance()
	for node in _get_all_children(scene_node):
		if node.get_class() == "TextureButton" and "target_scene" in node:
			if node.target_scene != "":
				if not node.target_scene in _cache.keys() and \
						not node.target_scene in _resource_queue.pending.keys():
					_resource_queue.queue_resource(node.target_scene)
					_queued_items.append(node.target_scene)


func _get_all_children(node:Node)->Array:
	var nodes: Array
	for child in node.get_children():
		nodes.append(child)
		if child.get_child_count() > 0:
			nodes.append_array(_get_all_children(child))
	return nodes
```

### PR DESCRIPTION
**Pull request by ThmsKnz**
BREAKING CHANGE: Adjust cache properties in configuration.tres and generate a cache map before export

Fixes https://github.com/deep-entertainment/issues/issues/38

### CODE REVIEW

**Comment by ThmsKnz** [comment]
Note: The new caching has also been successfully tested by Mikael in his current project.

File: addons/egoventure/cache/cache_update_dialog.gd [hunk 1]
```diff
@@ -0,0 +1,261 @@
+tool
+class_name CacheUpdateDialog
+extends WindowDialog
+
+
+# list of scenes in scene directory
+var _scene_list: Array
+var _scene_list_delta: Array
```
Comment (dploeger) [inline addons/egoventure/cache/cache_update_dialog.gd:9]: Please add a doc comment here as well

File: addons/egoventure/cache/cache_update_dialog.gd [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+# Cache map which gets generated
+var _cache_map = CacheMap.new()
+var _cache_map_time_modified
```
Comment (dploeger) [inline addons/egoventure/cache/cache_update_dialog.gd]: Please add a doc comment here as well

**Comment by ThmsKnz** [comment]
I've removed the commented lines you noticed.

File: addons/egoventure/cache/cache_update_dialog.gd [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+		var file = File.new()
+		_cache_map_time_modified = file.get_modified_time("res://cache_map.tres")
+	else:
+		_cache_map_time_modified = 0
+	
+	# retrieve list of scenes in scene directory
+	_scene_list.clear()
+	_scene_list_delta.clear()
+	_read_scene_list(scene_dir)
+#	_scene_list = _read_scene_list(scene_dir)
```
Comment (dploeger) [inline addons/egoventure/cache/cache_update_dialog.gd]: Please remove commented out code

File: addons/egoventure/cache/cache_update_dialog.gd [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+		% [_scene_list.size(), _scene_list_delta.size()]
+	$VBox/ProgressBar.value = 0.0
+	self.popup_centered()
+	# resize popup to ensure that it fits to rendered nodes
+	self.set_size($VBox.get_rect().size+Vector2(20,20))
+
+
+# Start updating the cache map
+func _on_Run_pressed(_full: bool):
+	#Deactivate buttons
```
Comment (dploeger) [inline addons/egoventure/cache/cache_update_dialog.gd]:
```suggestion
	# Deactivate buttons
```

File: addons/egoventure/cache/cache_update_dialog.gd [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+	$VBox/HBoxContainer/RunDelta.set_disabled(true)
+	$VBox/HBoxContainer/RunFull.set_disabled(true)
+	$VBox/HBoxContainer/Cancel.set_disabled(true)
+	# verbose remains active and can be toggled
```
Comment (dploeger) [inline addons/egoventure/cache/cache_update_dialog.gd]:
```suggestion
	# "verbose" button remains active and can be toggled
```

**Comment by dploeger** [comment]
I tested the branch now on mobile as well and on my old iPad the cache goes over the magic marker of 2GB when running down the road left of the house. Is there something to configure for this not to happen or is something wrong?

File: addons/egoventure/cache/cache_update_dialog.gd [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+	
+	$VBox/SceneCount.text = "The project contains %s scenes. (%s of them were changed after last cache map update)" \
+		% [_scene_list.size(), _scene_list_delta.size()]
+	$VBox/ProgressBar.value = 0.0
+	self.popup_centered()
+	# resize popup to ensure that it fits to rendered nodes
+	self.set_size($VBox.get_rect().size+Vector2(20,20))
+
+
+# Start updating the cache map
```
Comment (dploeger) [inline addons/egoventure/cache/cache_update_dialog.gd:52]: Add a parameter documentation for the _full parameter. Also, don't use a "\_"-prefix for that parameter

File: addons/egoventure/cache/cache_update_dialog.gd [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+
+# Recursively get all scene filenames from directory and subdirectories
+#
+# ** Parameters **
+#
+# - path: directory path
+# 
+# ** Returns ** List of scene filenames
+#func _read_scene_list(path: String) -> Array:
```
Comment (dploeger) [inline addons/egoventure/cache/cache_update_dialog.gd]: Remove (or use) this commented out code

File: addons/egoventure/cache/cache_update_dialog.gd [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+			filename = dir.get_next()
+		dir.list_dir_end()
+	
+	else:
+		print("An error occurred when trying to open directory %s" % path)
+
+
+# This will scan the scene and will return a size estimate and the adjacent scenes
+# Array[0]: size estimate, Array[1]: array of adjacent scenes
+func _scan_scene(scene_node) -> Array:
```
Comment (dploeger) [inline addons/egoventure/cache/cache_update_dialog.gd]: Add a doc comment about the parameter and (if possible) denote a type

File: addons/egoventure/cache/cache_update_dialog.gd [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+
+
+# Recursive function to retrieve all child nodes
+#
+# ** Parameters **
+#
+# - node: starting node
+#
+# ** Returns ** list of all child nodes
+func _get_all_children(node:Node)->Array:
```
Comment (dploeger) [inline addons/egoventure/cache/cache_update_dialog.gd]:
```suggestion
func _get_all_children(node: Node)->Array:
```

File: addons/egoventure/cache/scene_cache.gd [hunk 10]
```diff
@@ -29,6 +26,31 @@ var _queued_items: Array = []
 # Do not remove these scenes from the cache
 var _permanent_cache: PoolStringArray = []
 
+# Cache map
```
Comment (dploeger) [inline addons/egoventure/cache/scene_cache.gd]: Better describe what the "Cache map" does or contains

File: addons/egoventure/cache/scene_cache.gd [hunk 11]
```diff
@@ -29,6 +26,31 @@ var _queued_items: Array = []
...
+var _cache_map: CacheMap
+
+# Cache sizes (in kilobyte) - current and maximum
```
Comment (dploeger) [inline addons/egoventure/cache/scene_cache.gd]: Please split the comment into two above every variable (background is that the API docs are generated from that)

File: addons/egoventure/cache/scene_cache.gd [hunk 12]
```diff
@@ -29,6 +26,31 @@ var _queued_items: Array = []
...
+var _cache_size: int
+var _cache_max_size: int
+
+# Inner class for parameters of cached scenes
+class CacheMgmt:
+	var age: int
```
Comment (dploeger) [inline addons/egoventure/cache/scene_cache.gd:45]: Add doc comments on this and the next variable

File: addons/egoventure/cache/scene_cache.gd [hunk 13]
```diff
@@ -95,84 +134,135 @@ func get_scene(path: String) -> PackedScene:
...
-# filename: The path and filename of the scene
-func _get_index_from_filename(filename: String) -> int:
-	filename = filename.get_file()
-	var result = _scene_index_regex.search(filename)
-	if result == null:
-		return -1
-	return int(result.get_string("index"))
+#
+# - scene: scene that is to be added to permanent cache
+func update_permanent_cache(scene):
```
Comment (dploeger) [inline addons/egoventure/cache/scene_cache.gd]: If possible, please denote a type for the parameter

File: addons/egoventure/cache/scene_cache.gd [hunk 14]
```diff
@@ -95,84 +134,135 @@ func get_scene(path: String) -> PackedScene:
...
+	_resource_queue.queue_resource(scene)
+	_queued_items.append(scene)
+
+
+# Remove scenes from cache
+#
+# ** Parameters **
+#
+# - age_remove: all scenes up to this age get removed
```
Comment (dploeger) [inline addons/egoventure/cache/scene_cache.gd:218]: Please denote a type for the parameter

**Review by dploeger — CHANGES_REQUESTED** [review]
Such great work! Thanks a lot. Only some smaller comments to fix left.

**Comment by ThmsKnz** [comment]
Thanks for the change requests. All have been completed. In addition I removed one log file message that I only added for testing which is not needed in the final version.

**Review by dploeger — APPROVED** [review]
Wondeful. Thanks! 🎁

**Pull request merged**